README.md

```diff
@@ -1,2 +1,111 @@
 # SurrogatesBase.jl
-Basically just a surrogate in disguise
+
+API for deterministic and stochastic surrogates.
+
+Given data $((x_1, y_1), \ldots, (x_N, y_N))$ obtained by evaluating a function $y_i =
+f(x_i)$ or sampling from a conditional probability density $p_{Y|X}(Y = y_i|X = x_i)$,
+a **deterministic surrogate** is a function $s(x)$ (e.g. a [radial basis function
+interpolator](https://en.wikipedia.org/wiki/Radial_basis_function_interpolation)) that
+uses the data to approximate $f$ or some statistic of $p_{Y|X}$ (e.g. the mean),
+whereas a **stochastic surrogate** is a stochastic process (e.g. a [Gaussian process
+approximation](https://en.wikipedia.org/wiki/Gaussian_process_approximations)) that uses
+the data to approximate $f$ or $p_{Y|X}$ *and* quantify the uncertainty of the
+approximation.
+
+## Deterministic Surrogates
+
+Deterministic surrogates `s` are subtypes of `SurrogatesBase.AbstractDeterministicSurrogate`, 
+which is a subtype of `Function`.
+The method `add_points!(s, xs, ys)` **must** be implemented and the surrogate **must** be
+[callable](https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects)
+`s(xs)`, where `xs` is a `Vector` of inputs and `ys` is a `Vector` of corresponding outputs.
+For single points `x` and `y`, call these methods via `add_points!(s, [x], [y])`
+and `s([x])`.
+
+If the surrogate `s` has tunable hyper-parameters, the methods
+`update_hyperparameters!(s, prior)` and `hyperparameters(s)` have to be implemented.
+
+### Example
+
+```julia
+using SurrogatesBase
+
+struct RBF{T} <: AbstractDeterministicSurrogate
+    scale::T
+    centers::Vector{T}
+    weights::Vector{T}
+end
+
+(rbf::RBF)(xs) = [rbf.weights' * exp.(-rbf.scale * (x .- rbf.centers).^2)
+                  for x in xs]
+
+function SurrogatesBase.add_points!(rbf::RBF, xs, ys)
+    # update rbf.weights
+    return rbf
+end
+
+SurrogatesBase.hyperparameters(rbf::RBF) = (rbf.scale,)
+
+function SurrogatesBase.update_hyperparameters!(rbf::RBF, prior)
```

Comment:
What does a prior mean wrt to deterministic surrogates?

Reply:
A prior parameter should contain information for updating hyperparamters, but not strictly in a probabilistic sense.
Based on the type of a prior, there could be more ways of performing an update. For example if we use a polynomial surrogate, we can pass the maximal degree and refit the model, or we could pass an array of possible degrees and implement a strategy for chooing the best one & refit the model.

Reply:
Even for deterministic surrogates one may want to set box constraints to define the allowed region. This can be seen as a very simple (uniform) prior.

Reply:
I am not sure I understand. What type of constaints?

Reply:
Constraints on hyperparameters in hyperparameter optimization. E.g., for a real valued hyperparameter `R`, restricting `R` to a lower $l$ and an upper bound $u$ can be seen as passing a uniform prior on $\left[ l,u \right]$  for `R`.

Reply:
E.g. the lower and upper bound of allowed values of the scale of a radial basis function.

Reply:
I understand we can pass in lower bound and upper bound. I feel it can be confusing to name it "prior". Maybe we pass in a tuple containing both explicitly?

Reply:
I would like to keep the name "prior". Passing lower and upper bounds really is just like a uniform prior over a interval/box. Even in the deterministic setting, one may want to favor some hyperparameter values over others and implement this with a special prior-like object. The interface should be flexible enough such that one can define different update functions
```julia
update_hyperparameters!(s::SomeSurrogate, prior::NamedTuple{(:lower, :upper)}) = ...
update_hyperparameters!(s::SomeSurrogate, prior::GaussianPrior) = ...
```
Maybe we should explain this in the docstring/README?

Reply:
I think it is better to not specify what type `prior` has, e.g. in an implementation of a Gaussian process surrogate that is currently under development and should eventually implement this interface, a `prior` is a struct containing both box constraints on hyperparameters and information specific to how to perform hyperparameter optimization. 

Maybe a different name, something like `update_hyperparameters_config` or just `config` would be more descriptive.

Another benefit of not specifying the type of `prior` is to make it possible to dispatch on different hyperparameter optimization methods. For example there can be a trivial one that just rewrites the hyperparameters passed in `prior` without optimizing and another more complicated one, as mentioned in the previous comment with polynomials.

Reply:
I think we should have the examples in README to be fully sketched out if its not too difficult - not just have comments on what they can be. This can useful for both users and developers to get an idea of the workflow.

Reply:
I think it would not result in more clarity in the docs. I would suggest to point developers to surrogates that implement the interface for more details. For the GP example, the following is (just) a wrapper on AbstractGPs that is implementing this interface, and I think it would be too complicated, in particular with hyperparameter optimization.

[https://github.com/JuliaBayesianOptimization/SurrogatesAbstractGPs.jl](https://github.com/JuliaBayesianOptimization/SurrogatesAbstractGPs.jl)

Reply:
I will add the link once we settle on the API & the GP surrogate is updated to fit the API.